src/Microsoft.VisualStudio.Services.Agent/SecretMasker.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.VisualStudio.Services.Agent
+{
+    [ServiceLocator(Default = typeof(SecretMasker))]
+    public interface ISecretMasker : IAgentService
+    {
+        IEnumerable<ISecretMask> Masks { get; }
+
+        string MaskSecrets(string input);
+
+        void Add(ISecretMask mask);
+    }
+
+    public class SecretMasker : AgentService, ISecretMasker
+    {
+        private const string MaskedString = "********";
+
+        private readonly List<ISecretMask> _secretMasks = new List<ISecretMask>();
```

Comment:
Also for the type of workloads in SecretMasker we may a lock that supports single writers and multiple readers as ReaderWriterLock.